src/lcd.c

```diff
@@ -0,0 +1,228 @@
+
+    lcd_command(LCD_SETDDRAMADDR | (col + handler.row_offsets[row]));
+}
+
+void    lcd_display_off(void)
+{
+    handler.displaycontrol &= ~LCD_DISPLAYON;
+    lcd_command(LCD_DISPLAYCONTROL | handler.displaycontrol);
+}
+
+void    lcd_display_on(void)
+{
+    handler.displaycontrol |= LCD_DISPLAYON;
+    lcd_command(LCD_DISPLAYCONTROL | handler.displaycontrol);
+}
+
+void    lcd_cursor_off(void)
+{
+    handler.displaycontrol &= ~LCD_CURSORON;
+    lcd_command(LCD_DISPLAYCONTROL | handler.displaycontrol);
+}
+
+void    lcd_cursor_on(void)
+{
+    handler.displaycontrol |= LCD_CURSORON;
+    lcd_command(LCD_DISPLAYCONTROL | handler.displaycontrol);
+}
+
+void    lcd_blink_off(void)
+{
+    handler.displaycontrol &= ~LCD_BLINKON;
+    lcd_command(LCD_DISPLAYCONTROL | handler.displaycontrol);
+}
+
+void    lcd_blink_on(void)
+{
+    handler.displaycontrol |= LCD_BLINKON;
+    lcd_command(LCD_DISPLAYCONTROL | handler.displaycontrol);
+}
+
+void    lcd_scroll_display_left(void)
+{
+    lcd_command(LCD_CURSORSHIFT | LCD_DISPLAYMOVE | LCD_MOVELEFT);
+}
+
+void    lcd_scroll_display_right(void)
+{
+    lcd_command(LCD_CURSORSHIFT | LCD_DISPLAYMOVE | LCD_MOVERIGHT);
+}
+
+void    lcd_left_to_right(void)
+{
+    handler.displaymode |= LCD_ENTRYLEFT;
+    lcd_command(LCD_ENTRYMODESET | handler.displaymode);
+}
+
+void    lcd_right_to_left(void)
+{
+    handler.displaymode &= ~LCD_ENTRYLEFT;
+    lcd_command(LCD_ENTRYMODESET | handler.displaymode);
+}
+
+void    lcd_autoscroll_on(void)
+{
+    handler.displaymode |= LCD_ENTRYSHIFTINCREMENT;
+    lcd_command(LCD_ENTRYMODESET | handler.displaymode);
+}
+
+void    lcd_autoscroll_off(void)
+{
+    handler.displaymode &= ~LCD_ENTRYSHIFTINCREMENT;
+    lcd_command(LCD_ENTRYMODESET | handler.displaymode);
+}
+
+void    lcd_create_char(uint8_t location, uint8_t charmap[])
+{
+    location &= 0x7; // we only have 8 locations 0-7
+    lcd_command(LCD_SETCGRAMADDR | (location << 3));
+    for (int i = 0; i < 8; i++)
+        lcd_write(charmap[i]);
+}
+
+void    lcd_command(uint8_t value)
+{
+    gpio_set(GPIOA, handler.rs_pin, 0);
+    write4bits(value >> 4);
+    write4bits(value);
+}
+
+void    lcd_write(char value)
+{
+    gpio_set(GPIOA, handler.rs_pin, 1);
+    write4bits(value >> 4);
+    write4bits(value);
+}
+
+void    lcd_init(void)
+{
+    handler.rs_pin = 9;
+    handler.enable_pin = 7;
+    handler.data_pins[0] = 5;
+    handler.data_pins[1] = 4;
+    handler.data_pins[2] = 10;
+    handler.data_pins[3] = 8;
+    handler.numlines = 1;
+    handler.row_offsets[0] = 0;
+    handler.row_offsets[1] = 64;
+    handler.row_offsets[2] = 16;
+    handler.row_offsets[3] = 64 + 16;
+    handler.displayfunction = LCD_4BITMODE | LCD_1LINE | LCD_5x8DOTS;
+
+    GPIO_config gpio_config = {
+        .pin = handler.rs_pin,
+        .mode = GPIO_MODE_OUTPUT,
+        .pull = GPIO_NO_PUPD,
+        .speed = GPIO_SPEED_LOW,
+    };
+    gpio_init(GPIOA, &gpio_config);
+    gpio_config.pin = handler.enable_pin;
```

Comment:
Do not reuse the config object to initialize 3 different GPIO ports. It's less code, but the reader needs to keep track of what's inside the gpio_config after each line. This can also lead to bugs if someone modifies the code (e.g., I want to pull-up the rs_pin—if I change the gpio_config above, _all_ of the pins will become pull-up).